fastcrypto-zkp/src/bn254/zk_login.rs

```diff
@@ -1,86 +1,144 @@
-        (
-            OAuthProvider::Twitch.get_config().0,
-            SupportedKeyClaim::Email.to_string(),
-        ),
-        "",
-    );
-    map
-});
+const MAX_HEADER_LEN: u16 = 500;
+const PACK_WIDTH: u16 = 248;
+const ISS: &str = "iss";
+const AUD: &str = "aud";
+const NUM_EXTRACTABLE_STRINGS: u8 = 5;
+const MAX_EXTRACTABLE_STR_LEN: u16 = 150;
+const MAX_EXTRACTABLE_STR_LEN_B64: u16 = 4 * (1 + MAX_EXTRACTABLE_STR_LEN / 3);
 
 /// Supported OAuth providers. Must contain "openid" in "scopes_supported"
 /// and "public" for "subject_types_supported" instead of "pairwise".
-#[derive(Debug)]
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
 pub enum OAuthProvider {
     /// See https://accounts.google.com/.well-known/openid-configuration
     Google,
     /// See https://id.twitch.tv/oauth2/.well-known/openid-configuration
     Twitch,
+    /// See https://www.facebook.com/.well-known/openid-configuration/
+    Facebook,
+}
+
+/// Struct that contains all the OAuth provider information. A list of them can
+/// be retrieved from the JWK endpoint (e.g. <https://www.googleapis.com/oauth2/v3/certs>)
+/// and published on the bulletin along with a trusted party's signature.
+// #[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash, Serialize, Deserialize)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OAuthProviderContent {
+    /// Key type parameter, https://datatracker.ietf.org/doc/html/rfc7517#section-4.1
+    pub kty: String,
+    /// RSA public exponent, https://datatracker.ietf.org/doc/html/rfc7517#section-9.3
+    pub e: String,
+    /// RSA modulus, https://datatracker.ietf.org/doc/html/rfc7517#section-9.3
+    pub n: String,
+    /// Algorithm parameter, https://datatracker.ietf.org/doc/html/rfc7517#section-4.4
+    pub alg: String,
+    /// Key ID that identifies a JWK in a JWK set, https://datatracker.ietf.org/doc/html/rfc7517#section-4.5
+    kid: String,
+}
+
+/// Reader struct to parse all fields.
+// #[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash, Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct OAuthProviderContentReader {
+    e: String,
+    n: String,
+    #[serde(rename = "use")]
+    my_use: String,
+    kid: String,
+    kty: String,
+    alg: String,
+}
+
+impl OAuthProviderContent {
+    /// Get the kid string.
+    pub fn kid(&self) -> &str {
+        &self.kid
+    }
+
+    /// Parse OAuthProviderContent from the reader struct.
+    pub fn from_reader(reader: OAuthProviderContentReader) -> FastCryptoResult<Self> {
+        if reader.alg != "RS256" || reader.my_use != "sig" || reader.kty != "RSA" {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        Ok(Self {
+            kty: reader.kty,
+            kid: reader.kid,
+            e: trim(reader.e),
+            n: trim(reader.n),
+            alg: reader.alg,
+        })
+    }
+
+    /// Parse OAuthProviderContent from the reader struct.
+    pub fn validate(&self) -> FastCryptoResult<()> {
```

Comment:
Where is this used? It seems like some validation is being done in `from_reader` and some here. Can we consolidate so that validation only happens at one place?